File: apps/mobile/src/lib/appState.ts
export const DEFAULT_INTERVAL_DAYS = 3;
export const DEFAULT_MESSAGE =
  'Oi, não consegui confirmar que estou bem no app "Estou Vivo!". ' +
  'Pode verificar se está tudo certo comigo?';

export type AliveStatus =
  | 'neverChecked'
  | 'safe'
  | 'warning'
  | 'critical'
  | 'expired';

export type EmergencyContact = { name: string; phone: string };

export type AppConfig = {
  intervalDays: number;
  emergencyMessage: string;
  lastCheckIn: Date | null;
};

export function deadlineOf(c: AppConfig): Date | null {
  if (!c.lastCheckIn) return null;
  const d = new Date(c.lastCheckIn);
  d.setDate(d.getDate() + c.intervalDays);
  return d;
}

export function timeRemaining(c: AppConfig): Duration | null {
  const dl = deadlineOf(c);
  if (!dl) return null;
  const ms = dl.getTime() - Date.now();
  return ms <= 0 ? { ms: 0 } : { ms };
}

export type Duration = { ms: number };

export function isExpired(c: AppConfig): boolean {
  const tr = timeRemaining(c);
  return tr !== null && tr.ms === 0;
}

export function progressPercent(c: AppConfig): number {
  if (!c.lastCheckIn) return 0;
  const dl = deadlineOf(c);
  if (!dl) return 0;
  const total = dl.getTime() - c.lastCheckIn.getTime();
  if (total <= 0) return 1;
  const elapsed = Date.now() - c.lastCheckIn.getTime();
  return Math.min(1, Math.max(0, elapsed / total));
}

export function aliveStatus(c: AppConfig): AliveStatus {
  if (!c.lastCheckIn) return 'neverChecked';
  if (isExpired(c)) return 'expired';
  const tr = timeRemaining(c);
  if (!tr) return 'neverChecked';
  const h = tr.ms / 3600000;
  if (h <= 6) return 'critical';
  if (h <= 24) return 'warning';
  return 'safe';
}

File: legacy/flutter-estou-vivo/lib/models/app_state.dart
class AppConfig {
  static const int defaultIntervalDays = 3;
  static const String defaultMessage =
      'Oi, não consegui confirmar que estou bem no app "Estou Vivo!". '
      'Pode verificar se está tudo certo comigo?';

  int intervalDays;
  String emergencyMessage;
  DateTime? lastCheckIn;

  AppConfig({
    this.intervalDays = defaultIntervalDays,
    this.emergencyMessage = defaultMessage,
    this.lastCheckIn,
  });

  /// Calcula o prazo final (deadline) a partir do último check-in
  DateTime? get deadline =>
      lastCheckIn?.add(Duration(days: intervalDays));

  /// Calcula o tempo restante até o deadline
  Duration? get timeRemaining {
    if (deadline == null) return null;
    final remaining = deadline!.difference(DateTime.now());
    return remaining.isNegative ? Duration.zero : remaining;
  }

  /// Verifica se o prazo expirou
  bool get isExpired {
    if (deadline == null) return false;
    return DateTime.now().isAfter(deadline!);
  }

  /// Porcentagem do tempo consumido (0.0 a 1.0)
  double get progressPercent {
    if (lastCheckIn == null || deadline == null) return 0.0;
    final total = deadline!.difference(lastCheckIn!).inSeconds;
    final elapsed = DateTime.now().difference(lastCheckIn!).inSeconds;
    return (elapsed / total).clamp(0.0, 1.0);
  }
}
